src/sp800_108.rs

```diff
@@ -0,0 +1,386 @@
+            }
+            Sp800Params::ByteArray(param) => {
+                op.sign_update(param.as_slice())?;
+            }
+            Sp800Params::DKMLength(param) => {
+                if seen_dkmlen {
+                    return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+                }
+                seen_dkmlen = true;
+                dkm_update(param, dkmklen, dkmslen, op)?;
+            }
+        }
+    }
+
+    if !seen_iter {
+        return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+    }
+
+    Ok(())
+}
+
+fn feedback_updates(
+    params: &Vec<Sp800Params>,
+    op: &mut Box<dyn Sign>,
+    iv: &[u8],
+    ctr: usize,
+    dkmklen: usize,
+    dkmslen: usize,
+) -> KResult<()> {
+    let mut seen_dkmlen = false;
+    let mut seen_iter = false;
+    let mut seen_counter = false;
+    for p in params {
+        match p {
+            Sp800Params::Iteration(param) => {
+                if seen_iter {
+                    return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+                }
+                if param.defined {
+                    return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+                }
+                seen_iter = true;
+                op.sign_update(iv)?;
+            }
+            Sp800Params::Counter(param) => {
+                if seen_counter {
+                    return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+                }
+                seen_counter = true;
+                ctr_update(param, ctr, op)?;
+            }
+            Sp800Params::ByteArray(param) => {
+                op.sign_update(param.as_slice())?;
+            }
+            Sp800Params::DKMLength(param) => {
+                if seen_dkmlen {
+                    return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+                }
+                seen_dkmlen = true;
+                dkm_update(param, dkmklen, dkmslen, op)?;
+            }
+        }
+    }
+
+    if !seen_iter {
+        return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+    }
+
+    Ok(())
+}
+
+impl Derive for Sp800Operation {
+    fn derive(
+        &mut self,
+        key: &Object,
+        template: &[CK_ATTRIBUTE],
+        mechanisms: &Mechanisms,
+        objfactories: &ObjectFactories,
+    ) -> KResult<(Object, usize)> {
+        Self::verify_prf_key(self.prf, key)?;
+
+        /* Ok so this stuff in the PKCS#11 spec has an insane level
+         * of flexibility, fundamentally each parameter correspond to
+         * data that will be feed to the MAC operation in the order
+         * it should happen, providing maximum composability and an
+         * effectively infinite combinatorial matrix.
+         *
+         * This is an attempt at supporting insanity :-) */
+
+        let mechanism = CK_MECHANISM {
+            mechanism: self.prf,
+            pParameter: std::ptr::null_mut(),
+            ulParameterLen: 0,
+        };
+        let mech = mechanisms.get(self.prf)?;
+        let mut op = mech.sign_new(&mechanism, key)?;
```

Comment:
Slightly related, but RSA/EC keys have `key.check_key_ops` to check the key can be used for a given operation in `mech.sign_new()`, but HMAC keys do not have this. Should the check be present verifying the HMAC key can be used for signature and have `CKA_SIGN` attribute? If so, would it mean that the key used for derive here would need to be usable also for signing or given that this is done internally, the check should be skipped?